tests/ship_client.cpp

```diff
@@ -1,120 +1,159 @@
+
+  unixs::socket unix_socket(ctx);
+  ws::stream<unixs::socket> unix_stream(ctx);
+
+  bpo::options_description cli("ship_client command line options");
+  bool help = false;
+  std::string socket_address = "127.0.0.1:8080";
+  uint32_t num_requests = 1;
+
+  cli.add_options()("help,h", bpo::bool_switch(&help)->default_value(false),
+                    "Print this help message and exit.")(
+      "socket-address,a",
+      bpo::value<std::string>(&socket_address)->default_value(socket_address),
+      "Websocket address and port.")(
+      "num-requests,n",
+      bpo::value<uint32_t>(&num_requests)->default_value(num_requests),
+      "number of requests to make");
+  bpo::variables_map varmap;
+  bpo::store(bpo::parse_command_line(argc, argv, cli), varmap);
+  bpo::notify(varmap);
+
+  if (help) {
+    cli.print(std::cout);
+    return 0;
+  }
+
+  std::string statehistory_server, statehistory_port;
+
+  // unix socket
+  if (boost::algorithm::starts_with(socket_address, "ws+unix://") ||
+      boost::algorithm::starts_with(socket_address, "unix://")) {
+    statehistory_port = "";
+    statehistory_server = socket_address.substr(socket_address.find("unix://") +
+                                                strlen("unix://") + 1);
+  } else {
+    std::string::size_type colon = socket_address.find(':');
+    eosio::check(colon != std::string::npos,
+                 "Missing ':' seperator in Websocket address and port");
+    statehistory_server = socket_address.substr(0, colon);
+    statehistory_port = socket_address.substr(colon + 1);
+  }
+
+  std::cerr << "[\n{\n   \"status\": \"construct\",\n   \"time\": "
+            << time(NULL) << "\n},\n";
+
+  try {
+    auto run = [&]<typename SocketStream>(SocketStream &stream) {
       {
-         boost::beast::flat_buffer abi_buffer;
-         stream.read(abi_buffer);
-         std::string abi_string((const char*)abi_buffer.data().data(), abi_buffer.data().size());
-         eosio::json_token_stream token_stream(abi_string.data());
-         eosio::abi_def abidef = eosio::from_json<eosio::abi_def>(token_stream);
-         eosio::convert(abidef, abi);
+        boost::beast::flat_buffer abi_buffer;
+        stream.read(abi_buffer);
+        std::string abi_string((const char *)abi_buffer.data().data(),
+                               abi_buffer.data().size());
+        eosio::json_token_stream token_stream(abi_string.data());
+        eosio::abi_def abidef = eosio::from_json<eosio::abi_def>(token_stream);
+        eosio::convert(abidef, abi);
       }
 
-      std::cerr << "{\n   \"status\": \"set_abi\",\n   \"time\": " << time(NULL) << "\n},\n";
+      std::cerr << "{\n   \"status\": \"set_abi\",\n   \"time\": " << time(NULL)
+                << "\n},\n";
 
-      const eosio::abi_type& request_type = abi.abi_types.at("request");
-      const eosio::abi_type& result_type = abi.abi_types.at("result");
+      const eosio::abi_type &request_type = abi.abi_types.at("request");
```

Comment:
This codebase uses the convention of attaching type qualifiers to the type name, not the variable name.

Reply:
Not sure what should be done here, we have similar code in ship_streamer as well

Reply:
`eosio::abi_type& request_type` rather than `eosio::abi_type &request_type`.  I personally am most familiar/comfortable with the latter since I started out writing C which used that convention for pointers, but this repository uses the former.

Reply:
Ah, I see, got it! I thought something is wrong with the function used :) I like it more the C way as well. I screwed up formatting of this file, so I did clang-format it, and it does it...  Fixed.